### PR TITLE
[FIXED JENKINS-36537] Allow the use of custom json signature validator for metadata signature check

### PR DESCRIPTION
**Pull request by varmenise**
@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 1]
```diff
@@ -245,7 +245,16 @@ private FormValidation verifySignature(JSONObject o) throws IOException {
...
-        return new JSONSignatureValidator("update site '"+id+"'");
+        return getJsonSignatureValidator("update site '"+id+"'");
+    }
+
+    /**
+     * Let sub-classes of UpdateSite provide their own signature validator.
+     * @return the signature validator.
+     */
+    @Nonnull
+    protected JSONSignatureValidator getJsonSignatureValidator(String name) {
```
Comment (jtnord) [inline core/src/main/java/hudson/model/UpdateSite.java]: why can't subclasses override `getJsonSignatureValidator()`?

File: core/src/main/java/hudson/model/DownloadService.java [hunk 2]
```diff
@@ -324,6 +324,21 @@ public String getUrl() {
         }
 
         /**
+         * URL of specific updateSite.
+         */
+        public String getUrl(UpdateSite site) {
```
Comment (jtnord) [inline core/src/main/java/hudson/model/DownloadService.java]: would this not be better defined in UpdateSite so that subclasses can change the location?

Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/DownloadService.java]: +1. It's location in Download service smells like a hack and needs justification

File: core/src/main/java/hudson/model/DownloadService.java [hunk 3]
```diff
@@ -324,6 +324,21 @@ public String getUrl() {
         }
 
         /**
+         * URL of specific updateSite.
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/DownloadService.java]: The method may return null. it needs to be documented and annotated properly

File: core/src/main/java/hudson/model/DownloadService.java [hunk 4]
```diff
@@ -397,7 +412,8 @@ private FormValidation load(String json, long dataTimestamp) throws IOException
         public FormValidation updateNow() throws IOException {
             List<JSONObject> jsonList = new ArrayList<>();
             boolean toolInstallerMetadataExists = false;
-            for (String site : getUrls()) {
+            for (UpdateSite updatesite : Jenkins.getActiveInstance().getUpdateCenter().getSiteList()) {
+                String site = getUrl(updatesite);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/DownloadService.java]: 🐜  Unchecked null. Not fatal, but the error propagation is bad

File: core/src/main/java/hudson/model/UpdateSite.java (hunk 1 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/UpdateSite.java]: I would deprecate the non-parameterized method and make this one CheckForNull with a fallback to <code>id</code>. Anyway, the inheritance seems to be fragile, not sure why external users need to pass the "update site '" text prefix

Comment (varmenise) [inline core/src/main/java/hudson/model/UpdateSite.java]: cause the signature Validator can also be called by Download Service and in that case the name would be "downloadable" according to the current implementation. I ll fix the deprecation

File: core/src/main/java/hudson/model/DownloadService.java (hunk 2 above)
Comment (varmenise) [inline core/src/main/java/hudson/model/DownloadService.java]: well, this is the URL for the update Site and for a specific download service, I can move it to UpdateSite and send the DownloadService id I want to retrieve the URL for, or keep it here and pass the UpdateSite as parameter..I will go with your suggestion although they are inter-dependent and I do not see a big difference.

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 5]
```diff
@@ -422,6 +436,26 @@ public String getUrl() {
         return url;
     }
 
+
+    /**
+     * URL which exposes the metadata location in a specific update site.
+     * @param downloadable, the downloadable id of a specific metatadata json (e.g. hudson.tasks.Maven.MavenInstaller.json)
+     * @return the location
+     */
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/UpdateSite.java]: NIT: <code>@since</code>

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 6]
```diff
@@ -243,9 +243,23 @@ private FormValidation verifySignature(JSONObject o) throws IOException {
...
+
+    /**
+     * Let sub-classes of UpdateSite provide their own signature validator.
+     * @param name, the name for the JSON signature Validator object
+     * @return the signature validator.
+     */
+    @Nonnull
+    protected JSONSignatureValidator getJsonSignatureValidator(@CheckForNull String name) {
+        if (name == null) {
+            return getJsonSignatureValidator();
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/UpdateSite.java]: Strange passing of calls between two methods. I'd rather add the default name specification here and then just pass null from non-parameterized getJsonSignatureValidator

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 7]
```diff
@@ -243,9 +243,23 @@ private FormValidation verifySignature(JSONObject o) throws IOException {
...
+    @Deprecated
     @Nonnull
     protected JSONSignatureValidator getJsonSignatureValidator() {
-        return new JSONSignatureValidator("update site '"+id+"'");
+        return getJsonSignatureValidator("update site '"+id+"'");
+    }
+
+    /**
+     * Let sub-classes of UpdateSite provide their own signature validator.
+     * @param name, the name for the JSON signature Validator object
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/UpdateSite.java]: Nice2have: describe the default behavior

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 8]
```diff
@@ -243,9 +243,23 @@ private FormValidation verifySignature(JSONObject o) throws IOException {
      * Let sub-classes of UpdateSite provide their own signature validator.
      * @return the signature validator.
      */
+    @Deprecated
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/UpdateSite.java]: Also add <code>@deprecated</code> Javadoc with a pointer to the replacement method

**Comment by oleg-nenashev** [comment]
@varmenise This fix changes the API of the core and may not be backported? It could be worked around by Restricted annotations, but I'm not 100% it's what you want. Do you need backporting?

**Comment by varmenise** [comment]
@oleg-nenashev do not think we need back-porting, this is done to fix a regression of latest versions anyway ...wdyt?

**Comment by oleg-nenashev** [comment]
> Plus, isn't this change safe since I am deprecating the old method which is still there?

It's not safe, because you add new protected and public methods to API. If the PR gets backported, it will extend API compared to 2.8. It's not acceptable for PRs, since it causes some risks for plugin implementations

**Comment by daniel-beck** [comment]
@oleg-nenashev Don't we generally `@Restricted` backported API additions? CC @olivergondza

File: core/src/main/java/hudson/model/DownloadService.java [hunk 9]
```diff
@@ -397,7 +397,11 @@ private FormValidation load(String json, long dataTimestamp) throws IOException
         public FormValidation updateNow() throws IOException {
             List<JSONObject> jsonList = new ArrayList<>();
             boolean toolInstallerMetadataExists = false;
-            for (String site : getUrls()) {
+            for (UpdateSite updatesite : Jenkins.getActiveInstance().getUpdateCenter().getSiteList()) {
+                String site = updatesite.getMetadataUrlForDownloadable(url);
+                if (site == null) {
+                    return FormValidation.warning("One of the update site does not look like an update center");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/DownloadService.java]: Grammar: `updates sites`

Also, should probably say which one (URL) it is.

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 10]
```diff
@@ -242,10 +242,28 @@ private FormValidation verifySignature(JSONObject o) throws IOException {
...
+     * @param name, the name for the JSON signature Validator object.
+     *              if name is null, then the default name will be used,
+     *              which is "update site" followed by the update site id
+     * @return the signature validator.
+     * @since 2.15
+     */
+    @Nonnull
+    protected JSONSignatureValidator getJsonSignatureValidator(@CheckForNull String name) {
+        if (name == null) {
+            name = "update site '" + id + "'";
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/UpdateSite.java]: This string seems like magic that has a tendency to be copied and pasted all over the place. Is there a better solution?

**Comment by varmenise** [comment]
@daniel-beck @oleg-nenashev if I make them @Restricted, will this be backportable?

**Comment by oleg-nenashev** [comment]
@varmenise IMHO yes. Commonly we add commit with Restricted to master and then a follow-up PR which removes restrictions (if we really want to remove them). In such case the original change can be easily backported

File: core/src/main/java/hudson/model/DownloadService.java (hunk 9 above)
Comment (jtnord) [inline core/src/main/java/hudson/model/DownloadService.java]: >Grammar: updates sites

@daniel-beck that smells wrong to me...   `update sites` (the singular being `update site` not `updates site` even if the singular as used thoughout Jenkins would potentially be wrong)

Comment (daniel-beck) [inline core/src/main/java/hudson/model/DownloadService.java]: @jtnord Yep, typo when pointing out a typo 😢

**Comment by varmenise** [comment]
@oleg-nenashev done is it good now?

**Comment by varmenise** [comment]
investigating failures...

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 11]
```diff
@@ -422,6 +445,28 @@ public String getUrl() {
         return url;
     }
 
+
+    /**
+     * URL which exposes the metadata location in a specific update site.
+     * @param downloadable, the downloadable id of a specific metatadata json (e.g. hudson.tasks.Maven.MavenInstaller.json)
+     * @return the location
+     * @since 2.15
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/UpdateSite.java]: To a person who merges the PR: update the version to an actual one

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 12]
```diff
@@ -242,10 +246,29 @@ private FormValidation verifySignature(JSONObject o) throws IOException {
...
+        return getJsonSignatureValidator(null);
+    }
+
+    /**
+     * Let sub-classes of UpdateSite provide their own signature validator.
+     * @param name, the name for the JSON signature Validator object.
+     *              if name is null, then the default name will be used,
+     *              which is "update site" followed by the update site id
+     * @return the signature validator.
+     * @since 2.15
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/UpdateSite.java]: same

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 13]
```diff
@@ -242,10 +246,29 @@ private FormValidation verifySignature(JSONObject o) throws IOException {
...
+     */
+    @Nonnull
+    @Restricted(NoExternalUse.class)
+    protected JSONSignatureValidator getJsonSignatureValidator(@CheckForNull String name) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/UpdateSite.java]: Restricted will prohibit usage of this method in plugins/libs with BCT. Is it intentional?

**Comment by oleg-nenashev** [comment]
The code looks good to me. Added minor comments

**Comment by oleg-nenashev** [comment]
I think the PR is good to be merged

**Comment by ndeloof** [comment]
🐝

**Comment by oleg-nenashev** [comment]
@reviewbybees done

**Comment by oleg-nenashev** [comment]
Merging as is. I'll follow-up on wrong <code>@since</code> definitions as a part of the bulk cleanup

File: core/src/main/java/hudson/model/UpdateSite.java (hunk 13 above)
Comment (varmenise) [inline core/src/main/java/hudson/model/UpdateSite.java]: no, this has to be overrdable. I am creating a follow up PR

**Pull request merged**